lib/route/neigh.c

```diff
@@ -762,6 +777,9 @@ static int build_neigh_msg(struct rtnl_neigh *tmpl, int cmd, int flags,
 	if (tmpl->ce_mask & NEIGH_ATTR_NHID)
 		NLA_PUT_U32(msg, NDA_NH_ID, tmpl->n_nhid);
 
+	if (tmpl->ce_mask & NEIGH_ATTR_EXT_FLAGS)
+		NLA_PUT_U32(msg, NDA_FLAGS_EXT, tmpl->n_ext_flags);
```

Comment:
kernel does:
```C
#define NTF_EXT_MASK      (NTF_EXT_MANAGED)
```
```C
[NDA_FLAGS_EXT]          = NLA_POLICY_MASK(NLA_U32, NTF_EXT_MASK),
```

Note that `NTF_EXT_LOCKED` is not part of the mask. That means, if we sent kernel a `NDA_FLAGS_EXT` attribute with  `NTF_EXT_LOCKED` flag set, it will be rejected.

This means, if we receive a neighbor instance from kernel, modify it slightly, and send it back, then it will be rejected.

As it never makes sense to send kernel such a flag (at least not today, until this restriction is lifted), it should be filtered out on the client side.

Also, if we don't have any flags to set, we maybe should not send the attribute. Older kernel versions would not know this attribute (though they may or may not ignore unknown netlink attributes to be forward compatible).


In summary, how about:
```C
    if (tmpl->ce_mask & NEIGH_ATTR_EXT_FLAGS) {
        uint32_t f = tmpl->n_ext_flags & (~((uint32_t) NTF_EXT_LOCKED));

        /* We cannot set NTF_EXT_LOCKED, filter it out.
         * Also, don't send the attribute if it's zero anyway, to be backward
         * compatible with older kernel. */
        if (f != 0)
           NLA_PUT_U32(msg, NDA_FLAGS_EXT, f);
    }
```

Reply:
> Also, if we don't have any flags to set, we maybe should not send the attribute. Older kernel versions would not know this attribute (though they may or may not ignore unknown netlink attributes to be forward compatible).

The code already has this behavior. The ext flags will only be sent to the kernel if `NEIGH_ATTR_EXT_FLAGS` is set in the `ce_mask`. This will only be set if the user has set or unset an ext flag themselves or the kernel sent a message with the `NDA_FLAGS_EXT` attribute.

I've added a commit to unset the locked flag in the request, and only send the attribute if the flags value is nonzero.

Should we also not allow the user to set the locked flag? That would require changing the function signature of the setter, unless a separate variable is stored which represents the user's desired setting for the locked flag (which is only used if set). In that case the kernel would report the error instead of libnl. These are just some ideas. Maybe that's overcomplicating things and the user should just not expect setting that flag to do anything (but not error either).

Reply:
> > Also, if we don't have any flags to set, we maybe should not send the attribute.
>
> The code already has this behavior. The ext flags will only be sent to the kernel if NEIGH_ATTR_EXT_FLAGS is set in the ce_mask.

Right. With "don't have any flags set", I meant that the flags are explicitly set to zero (with NEIGH_ATTR_EXT_FLAGS). Anyway.

> Should we also not allow the user to set the locked flag?

I think `rtnl_neigh_set_ext_flags()` should allow setting the flag (just as it does now). 
- It might make sense to record the LOCKED flag in a object that the user creates. 
- it's clear that libnl cannot possibly send this to kernel, so requiring the user to filter out the flag is not helpful.
- we set the flag when parsing it, so a user should also be able to create a similar object with the flag set.